mlrun/serving/routers.py

```diff
@@ -664,29 +701,17 @@ def _apply_logic(self, predictions: dict):
 
         :return: List of the resulting voted predictions
         """
-
-        # Flatten predictions by sample instead of by model as received
-        num_of_events = len(
-            [*predictions.values()][0]["outputs"][self.prediction_col_name]
-            if self.format_response_with_col_name_flag
-            else [*predictions.values()][0]["outputs"]
-        )
-        flattened_predictions = []
-        weights = []
-        for i in range(num_of_events):
-            event_pred = []
-            for model_name, v in predictions.items():
-                event_pred.append(
-                    v["outputs"][self.prediction_col_name][i]
-                    if self.format_response_with_col_name_flag
-                    else v["outputs"][i]
-                )
-                if i == 0:
-                    weights.append(self._weights[model_name])
-            flattened_predictions.append(event_pred)
-
+        flattened_predictions = np.array(
+            list(map(self._extract_pred, predictions.values()))
+        ).T
+        weights = [self._weights[model_name] for model_name in predictions.keys()]
         return self.logic(flattened_predictions, np.array(weights))
 
+    def _extract_pred(self, dictionary: dict):
+        if self.format_response_with_col_name_flag:
+            return dictionary["outputs"][self.prediction_col_name]
+        return dictionary["outputs"]
```

Comment:
**If `_extract_pred` is used elsewhere, you can ignore this comment**

I would have done it with lambda, take a look:

```suggestion
        flattened_predictions = np.array(
            list(
                map(
                    lambda dictionary: (
                        dictionary["outputs"][self.prediction_col_name]
                        if self.format_response_with_col_name_flag
                        else dictionary["outputs"]
                    ), 
                    predictions.values()
                )
            )
        ).T
        weights = [self._weights[model_name] for model_name in predictions.keys()]
        return self.logic(flattened_predictions, np.array(weights))
```

If you don't think so at least move it to be an inner function :)